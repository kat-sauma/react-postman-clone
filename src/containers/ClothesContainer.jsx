import React, { Component } from 'react'
import Controls from '../components/Controls';
import Display from '../components/Display';
import { httpCall, postToApi, putApi, deleteFromApi } from '../service/ApiCall';
import Spinner from '../components/Spinner';

export default class ClothesContainer extends Component {
state = {
    loading: false,
    url: '',
    method: 'get',
    results: {},
    json: ''
};

//handle on change
handleOnChange = ({ target }) => {
    this.setState({ url: target.value });
};

handleMethodChange = (e) => {
    this.setState({ method: e.target.value })
}

handleJsonChange = (e) => {
    this.setState({ json: e.target.value }, () => console.log(this.state.json))
}

// handle submit -- loading true and then false once loaded
handleSubmit = async (event) => {
    event.preventDefault();

    const { url } = this.state;
    this.setState({ loading: true });
    const clothes = await httpCall(url);

    if (this.state.method === 'get') {
        let results = clothes;
        this.setState({ results: results });
    } else if (this.state.method === 'post') {
            let results = await postToApi(url, json);
            this.setState({ results: results });
    } else if (this.state.method === 'put') {
            let results = await putApi(url, json);
            this.setState({ results: results });
    } else {
        let results = await deleteFromApi(url);
        this.setState({ results: results });
    }

    this.setState({ loading: false });
}

    render() {
        const { loading, url, results, json } = this.state;
        
        if (loading) return <Spinner />;

        return (
            <main className='organize-closet' aria-label='organize closet'>
                <Controls 
                    url={url}
                    json={json}
                    onUrlChange={this.handleOnChange}
                    onMethodChange={this.handleMethodChange}
                    onJsonChange={this.handleJsonChange}
                    onSubmit={this.handleSubmit}
                    
                />
                <Display results={results} />
            </main>
        )
    }
}
