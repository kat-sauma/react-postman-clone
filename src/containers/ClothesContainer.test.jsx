import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ClothesContainer from './ClothesContainer';


describe('Clothes Container', () => {
    it('renders the form that holds controls, the body of returned json, and the history list', async () => {

        render(<ClothesContainer />);
        
        const inputURL = await screen.findByRole('textbox', { name: 'url' })
        userEvent.type(url, 'https://yalla-habibi.herokuapp.com/clothes')
        expect(inputURL).toMatchSnapshot();

        const getMethod = screen.getByRole('radio', { name: 'get' })
        expect(getMethod).toMatchSnapshot();
        

        const button = await screen.findByRole('button', { name: 'find clothes' });
        expect(button).toMatchSnapshot();
        userEvent.type(url, 'https://yalla-habibi.herokuapp.com/clothes/1')


        userEvent.click(button);
        screen.getByAltText('loading-spinner');
    });
});