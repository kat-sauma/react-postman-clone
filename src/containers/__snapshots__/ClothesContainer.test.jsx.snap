// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 1`] = `
<input
  aria-label="url"
  id="url"
  placeholder="http://type-api-url-here.com"
  type="url"
  value="https://yalla-habibi.herokuapp.com/clothes"
/>
`;

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 2`] = `
<input
  checked=""
  id="get"
  name="method"
  type="radio"
  value="get"
/>
`;

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 3`] = `
<input
  id="post"
  name="method"
  type="radio"
  value="post"
/>
`;

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 4`] = `
<input
  id="put"
  name="method"
  type="radio"
  value="put"
/>
`;

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 5`] = `
<input
  id="delete"
  name="method"
  type="radio"
  value="delete"
/>
`;

exports[`Clothes Container renders the form that holds controls, the body of returned json, and the history list 6`] = `
<button
  aria-label="find clothes"
  class="find-clothes"
>
  Submit
</button>
`;
